Redefine the simdlen semantic:

EikanWang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88736
* __->__ #88482

This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

EikanWang committed:
Redefine the simdlen semantic:
None: Automatically determin the SIMD
<=1: Explicitly disable SIMD
>1: Expelicitly specify the SIMD bit width

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


None: Automatically determin the SIMD
<=1: Explicitly disable SIMD
>1: Expelicitly specify the SIMD bit width

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


- None: Automatically determin the SIMD bit width
- <=1: Explicitly disable SIMD
- <>1: Expelicitly specify the SIMD bit width

Regarding for the last case, we will disable the SIMD if the specified value is not equal to the ISA bitwidth.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


- None: Automatically determin the SIMD bit width
- <=1: Explicitly disable SIMD
- <>1: Expelicitly specify the SIMD bit width

Regarding for the last case, we will disable the SIMD if the specified value is not equal to the ISA bitwidth.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


- None: Automatically determin the SIMD bit width
- <=1: Explicitly disable SIMD
- <>1: Expelicitly specify the SIMD bit width

Regarding for the last case, we will disable the SIMD if the specified value is not equal to the ISA bitwidth.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


- None: Automatically determin the SIMD bit width
- <=1: Explicitly disable SIMD
- <>1: Expelicitly specify the SIMD bit width

Regarding for the last case, we will disable the SIMD if the specified value is not equal to the ISA bitwidth.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


- None: Automatically determin the SIMD bit width
- <=1: Explicitly disable SIMD
- <>1: Expelicitly specify the SIMD bit width

Regarding for the last case, we will disable the SIMD if the specified value is not equal to the ISA bitwidth.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


- None: Automatically determin the SIMD bit width
- <=1: Explicitly disable SIMD
- <>1: Expelicitly specify the SIMD bit width

Regarding for the last case, we will disable the SIMD if the specified value is not equal to the ISA bitwidth.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


- None: Automatically determin the SIMD bit width
- <=1: Explicitly disable SIMD
- <>1: Expelicitly specify the SIMD bit width

Regarding for the last case, we will disable the SIMD if the specified value is not equal to the ISA bitwidth.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,176 @@ def is_gcc():
-@functools.lru_cache(1)
-def get_cpu_proc_info():
+@functools.lru_cache(None)
+def valid_vec_isa():
```
`valid_vec_isa_list`?

EikanWang replied:
Done

jgong5 commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,176 @@ def is_gcc():
-            return ""
+
+invalid_vec_isa = InvalidVecISA()
+supported_vec_isa = [VecAVX512(), VecAVX2()]
```
`supported_vec_isa_list`?

EikanWang replied:
Done

jgong5 commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,176 @@ def is_gcc():
-        return isa_info
+def pick_vec_isa():
+    _valid_vec_isa: List[VecISA] = valid_vec_isa()
+    if not _valid_vec_isa:
+        return invalid_vec_isa
 
+    # If the simdlen is None, it indicates determin the vectroization length automatically
+    if config.cpp.simdlen is None:
+        assert _valid_vec_isa
+        return _valid_vec_isa[0]
```
Not all paths return value?

EikanWang replied:
Could not capture your point. Do you mean we need to add "else"?

jgong5 replied:
Yes, make sure all control paths return value.

EikanWang replied:
The `else` logic is the left logic as `config.cpp.simdlen` is `None` or a concrete number.

jgong5 commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,176 @@ def is_gcc():
-            isa_info.append(_SupportedVecIsa.AVX2)
 
-        return isa_info
+def pick_vec_isa():
```
`preferred_vec_isa` or maybe `current_vec_isa`?

EikanWang replied:
The style of the function name is similar with `pick_loop_order` https://github.com/pytorch/pytorch/blob/master/torch/_inductor/ir.py#LL1971C22-L1971C22

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1347,7 +1349,9 @@ class LoopLevel:
     steps: sympy.Expr = sympy.Integer(1)
     parallel: int = 0
     simd_omp: bool = False
-    simd_len: int = config.cpp.simdlen
+    simd_len: int = (
+        codecache.pick_vec_isa().nelements() if codecache.pick_vec_isa() else 0
+    )
```
maybe keep the `VecISA` object directly here would be simpler?

EikanWang replied:
Done

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1361,7 +1365,11 @@ def lines(self):
-        simd = f"simd simdlen({self.simd_len}) " if self.simd_omp else ""
+        simd = (
+            f"simd simdlen({self.simd_len}) "
+            if self.simd_omp and self.simd_len > 1
```
Check `nelements` instead?

EikanWang replied:
Good idea.

EikanWang replied:
Done

jgong5 commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,176 @@ def is_gcc():
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+    @functools.lru_cache(None)
```
I guess you don't have to do `lru_cache` here suppose the validity check would always be invoked once inside another lru_cached function?

jgong5 replied:
In that case, you would have to make the lru_cache instance local then since it is now applied to a class method. See this: https://rednafi.github.io/reflections/dont-wrap-instance-methods-with-functoolslru_cache-decorator-in-python.html

Alternatively, you may consider to do such a compilation check in a dedicated method which would only be invoked once in the validity check. Then in the subclasses, you would either do `def __bool__(self): return True` or `def __bool__(self): return False`. The `picked_vec_isa` would be either an invalid instance or an instance that has been validated working on the CPU.

EikanWang replied:
The `picked_vec_isa` always returns the global class instance. Hence it does not impact the performance and not cache too much things.

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,170 @@ def is_gcc():
+        from filelock import FileLock
+
+        lock_dir = get_lock_dir()
+        lock = FileLock(os.path.join(lock_dir, key + ".lock"), timeout=LOCK_TIMEOUT)
```
Could we reuse more of the standard compile path for this?  

It might also make sense to cache the result to disk similar to the triton autotuner.  I worry this will hurt compile times.

EikanWang replied:
Hi Janson, I encountered a performance issue If I use the standard compiler path to load the generated library to check ISA. The performance of MT5ForConditionalGeneration dropped from 4.01x to 1.77x for multi-thread. I haven't investigated the root cause as I'm focusing on optimizing the outlier models. Hence, this PR creates another process to dry-run a python script to load the generated library to isolate the execution environment.

I will file an issue to track your comments and polish it later. Are you okay with that?

jgong5 replied:
OK. In that case, `lru_cache(1)` is more appropriate?

EikanWang replied:
It always returns the same result. I'm also curious about it because it only impacts `MT5ForConditionalGeneration`.

EikanWang replied:
> OK. In that case, lru_cache(1) is more appropriate?

Hi @jgong5 , `lru_cache(1)` will evict the previous `VecISA` instance. Since there are three global `VecISA` instances, we should not use `lru_cache(1)` from the performance perspective.

jgong5 replied:
Ah, yes, you are right. Sorry about the confusion.

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,176 @@ def is_gcc():
+    _dtype_nelements: Dict[torch.dtype, int]
+
+    # TorchInductor CPU vectorization reuses PyTorch vectorization utility functions
+    # Hence, TorchInductor would depend on Slee* to accelerate mathematical functions
```
Done

EikanWang commented on torch/_inductor/codecache.py:
```diff
@@ -147,79 +147,176 @@ def is_gcc():
+    # avx512 check in CMake - FindAVX.cmake. But TorchInductor install the latest
+    # gcc/g++ compiler by default while it could support the AVX512 compilation.
+    # Therefore, there would be a conflict sleef version between PyTorch and
+    # TorchInductor. Hence, we dry-run the following code to check whether current
```
Done

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -996,12 +997,49 @@ def codegen_loops(self, code, worksharing):
+        #    LINE07:          auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + (8*i0) + (16*i2) + (32*i1));
+        #    LINE08:          tmp1_vec += tmp0;
+        #    LINE09:      }
+        #    LINE10:      tmp1 = vec_reduce_all<float>([](Vectorized<float>& x, Vectorized<float>&y) {return x + y;}, tmp1_vec);
```
Alternatively, we can add a condition `extend > 0` (here, it is `0 > 0`) around this statement so that you don't have to specially handle the small range case in the codegen.

EikanWang replied:
I considered this design and decided to deprioritize it first as it is more complicated than the current design. Because we could not get the reduction range during the reduction suffix generation while `LINE10` is a reduction suffix. As I mentioned before, loop body generation is independent of loop generation.

I think we should have a more solid design to handle similar cases.

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Redefine the simdlen semantic:"


This PR is targeting to automatically enable vectorization optimization for TorchInductor. It refined the semantics of `config.cpp.simdlen`. 

Originally, `None` means to disable vectorization while a specific value means the number of elements to be vectorized once time. But it depends on the data. Regarding 256bit SVE/SIMD ISA for ARM and X86, the `simdlen` should be 16 for Float while 32 for BFloat. Hence, this PR defined the `simdlen` as the bit width. The detailed semantics are as follows.

- **_simdlen = None_**: Automatically determine the SIMD bit width. Detect HW information and pick the proper vectorization ISA. Specific for X86, the priority of AVX512 is higher than AVX2.
- **_simdlen <=1_**: Explicitly disable SIMD
- **_simdlen > 1_**: Explicitly specify the SIMD bit width. It equals the disabled semantic if the bit width does not match the ISA width.



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "Broke multiple tests in several trunk workflows, for example https://github.com/pytorch/pytorch/actions/runs/3485086792/jobs/5830429554" -c weird
